doc/parameter_server.md

```diff
@@ -0,0 +1,198 @@
+5. Pserver saves model checkpoint to disk periodially
+
+
+**Note**
+
+We could implement the kv store by ourselves, or we could use some already solution, such as redis.
+
+## Worker
+
+- Worker defines a forward/backward computation proccess
+- Worker defines a dataloader module
+
+Workflow:
+
+1. Before forward layer starting, worker sends key to pserver, and pull <key, parameter> back from pserver
+2. Worker starts forward layer computation
+3. After backward layer finishing, worker generates <key, gradient>
+4. Worker pushes <key, gradient> to pserver
+
+
+There is also another kind of worker who does an evalution job. It define its own evalution computation process and dataloader of validation data. It pulls <key, parameter> from pserver, but never push data to pserver.
+
+
+**Note**
+
+There could be some same item id in a minibatch data. So some gradient vector of embedding table will have the same item id. We need to sum these gradient before pushing to pserver.
+
+
+## Delayed model updating
+
+In order to reduce the communication overhead between workers and pservers, we propose a stragety called delayed model updating.
+
+A worker runs a round of forward/backward computation using its local model, and keep the gradients in its local. After several rounds finishing, it push gradients to pservers.
+
+**Note**
+
+Since local model is only a part of global model, in some circumstance, worker still has to pull embedding vector parameter from pservers if there exits unknow item ids in a minibatch data.
+
+In async mode, this will lead to relative newer embedding part parameters, but relative older other part parameters.
+
+
+## Elastic scheduling
+
+### Analysis
+
+First, we analyse the hardware resources that pserver and worker consume.
+
+
+| role| CPU | memory | network bandwidth |  disk |
+| :----  |:----  |:---|:---|:---|
+| worker  | high| medium |  medium | low|
+| pserver | low | medium | high| medium |
+
+
+Worker does the forward/backward computation and consumes CPU most, while pserver provides kvstore service and consumes network bandwidth most.
+
+Since our target is to scheduling deep learning jobs elasticly, we must have the ability to increase/decrease CPU resources of worker, and increase/decrease network bandwidth resources of pserver.
+
+### Solution
+
+As we are using asynchronous SGD, increasing/decreasing CPU resources of worker is trivial. We can just create more workers or kill some workers under different cluster status.
+
+And for pserver, how can we adjust network bandwidth? One solution is to create or delete pserver nodes. Since each node has a network interface card, more pserver nodes means more network bandwidth.
+
+The second solution is to adjust network bandwidth of current pserver node. We can create many pserver nodes first, but set network bandwidth limit to certain medium value. If we want to increase/decrease network bandwith, we increase/decrease the network bandwidth limit.
```

Comment:
ElasticDL should not change network bandwidth. 
Network bandwidth is very complicated in a data center network environment. For different k8s pods, they may reside in a physical machine, and share PCIe and network cards bandwidth. Or in different machines connected by routers/switches. From pods, there is no way to tell what network resources are shared and if the corresponding network bandwidth already exceeds its limit or not. Decrease a pod's network bandwidth may have no effect on any other ElasticDL pods, but decrease performance of itself.

Reply:
Yes, network environment could be complicated in a data center.
In an EDL job, workers need CPU most, pservers need network bandwidth most. Assigning CPU resources to workers is trivial, assigning network bandwidth resources to pservers is hard.

So, here comes some questions:

- Do we need to schedule pservers elastically? Or we just schedule workers.
- If we have to schedule pservers, how could we allocate the network bandwidth resources to different EDL jobs? 
- If EDL jobs are running with other jobs, how could we divide network bandwidth between EDL jobs and other jobs? We know dividing CPU resources is trivial.